### PR TITLE
Support Llama3 8b/70b

### PR DESCRIPTION
**Pull request by wanchaol**
This PR adds support for Llama3 8b/70b, mainly it:
- add tiktonizer, add instructions to download tokenizer
- add options for the llama model to support Llama3
- add Llama3 8b/70b configs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,15 +47,27 @@ Install PyTorch from source or install the latest pytorch nightly, then install
 pip install -r requirements.txt
 ```
 
-Install additional dev requirements if you want to contribute to the repo:
+### Downloading a tokenizer.model
+
+`torchtitan` current supports training Llama3 (8B, 70B), and Llama2 (13B, 70B) out of box. To get started training these models, we need to download a tokenizer.model. Follow the instructions on the official [meta-llama](https://huggingface.co/meta-llama/Meta-Llama-3-8B) repository to ensure you have access to the Llama model weights.
```
Comment (lessw2020) [inline README.md]: `torchtitan` current**ly** supports training...out of **_the_** box.

File: README.md [hunk 2]
```diff
@@ -47,15 +47,27 @@ Install PyTorch from source or install the latest pytorch nightly, then install
...
+Once you have confirmed access, you can run the following command to download the Llama2/3 tokenizer to your local machine.
+
 ```
-pip install -r dev-requirements.txt
+# pass your hf_token in order to download tokenizer.model
+
+# llama3 tokenizer.model
+python torchtitan/datasets/download_tokenizer.py --repo_id meta-llama/Meta-Llama-3-8B --local_dir "original" --hf_token=...
+
+# llama2 tokenier.model
```
Comment (lessw2020) [inline README.md]: tokeni**z**er.model

File: requirements.txt [hunk 3]
```diff
@@ -1,5 +1,6 @@
 torch >= 2.2.0.dev
-sentencepiece
 datasets
 tomli >= 1.1.0 ; python_version < "3.11"
 tensorboard
+sentencepiece
+tiktoken==0.4.0
```
Comment (lessw2020) [inline requirements.txt]: just curious - pip install is currently 0.6.0.  Is there a need to freeze on exactly 0.4.0?

File: torchtitan/datasets/tokenizer/base_tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import os
+
+from abc import ABC, abstractmethod
+from typing import List
+
+
+class TokenizerIf(ABC):
```
Comment (lessw2020) [inline torchtitan/datasets/tokenizer/base_tokenizer.py]: minor nit - TokenizerIf sounds like 'it might be a tokenizer or might not'.  The real meaning is Interface, but 'If' is an actual word so it reads poorly vs ideally should be intuitive from the name alone.
Might be easier to call it just TokenizerABC, Tokenizer, TokenizerBase, or else TokenizerInterface to be clearer.

File: README.md [hunk 5]
```diff
@@ -47,15 +47,27 @@ Install PyTorch from source or install the latest pytorch nightly, then install
...
 
-Install additional dev requirements if you want to contribute to the repo:
+### Downloading a tokenizer.model
+
+`torchtitan` current supports training Llama3 (8B, 70B), and Llama2 (13B, 70B) out of box. To get started training these models, we need to download a tokenizer.model. Follow the instructions on the official [meta-llama](https://huggingface.co/meta-llama/Meta-Llama-3-8B) repository to ensure you have access to the Llama model weights.
+
+Once you have confirmed access, you can run the following command to download the Llama2/3 tokenizer to your local machine.
+
 ```
-pip install -r dev-requirements.txt
```
Comment (tianyu-l) [inline README.md:52]: Was this removed intentionally?

File: README.md [hunk 6]
```diff
@@ -47,15 +47,27 @@ Install PyTorch from source or install the latest pytorch nightly, then install
...
+# pass your hf_token in order to download tokenizer.model
+
+# llama3 tokenizer.model
+python torchtitan/datasets/download_tokenizer.py --repo_id meta-llama/Meta-Llama-3-8B --local_dir "original" --hf_token=...
+
+# llama2 tokenier.model
+python torchtitan/datasets/download_tokenizer.py --repo_id meta-llama/Llama-2-13b-hf --hf_token=...
+
```
Comment (tianyu-l) [inline README.md]: nit: remove this last blank line

File: torchtitan/datasets/tokenizer/base_tokenizer.py (hunk 4 above)
Comment (tianyu-l) [reply torchtitan/datasets/tokenizer/base_tokenizer.py]: Agreed .. I was always confused when first looking at the name. IMO `Tokenizer` is good.

File: torchtitan/datasets/tokenizer/tiktokenizer.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    Union,
+)
+
+import tiktoken
+from tiktoken.load import load_tiktoken_bpe
+
+from torchtitan.datasets.tokenizer.base_tokenizer import TokenizerIf
+
+
+logger = getLogger(__name__)
```
Comment (tianyu-l) [inline torchtitan/datasets/tokenizer/tiktokenizer.py]: Currently for all other files we are doing `from torchtitan.logging_utils import logger`.
The difference is that it is a logger at the "root" level. Let's keep them consistent, either way is fine to me. The easiest is to follow other files, we can refactor later if needed.

File: torchtitan/models/__init__.py [hunk 8]
```diff
@@ -4,16 +4,16 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
-from torchtitan.models.llama import llama_configs, Transformer
+from torchtitan.models.llama import llama2_configs, llama3_configs, Transformer
 
 models_config = {
-    "llama": llama_configs,
+    "llama": llama2_configs,
```
Comment (tianyu-l) [inline torchtitan/models/__init__.py]: shall we rename llama to llama2, to deliver the message that it is not llama1?
we can do it in another PR.

File: train_configs/llama3_8b.toml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+name = "AdamW"
+lr = 3e-4
+
+[training]
+batch_size = 2
+seq_len = 8192
+warmup_steps = 200  # lr scheduler warm up
+max_norm = 1.0  # grad norm clipping
+steps = 1000
+# only dp would be sufficient for 7B
```
Comment (tianyu-l) [inline train_configs/llama3_8b.toml]: comment needs to be modified

File: train_configs/llama3_8b.toml [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+data_parallel_degree = -1
+tensor_parallel_degree = 1
+pipeline_parallel_degree = 1
+fp8_linear = ""
+compile = false
+dataset = "c4_mini"
```
Comment (tianyu-l) [inline train_configs/llama3_8b.toml]: default should be c4?

File: train_configs/llama3_8b.toml [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+[checkpoint]
+interval = 500
+interval_type = "steps"
+folder = ""
+model_weights_only = false
+
+[activation_checkpoint]
+mode="full"  # mode = 'selective'  # ['none', 'selective', 'full']
```
Comment (tianyu-l) [inline train_configs/llama3_8b.toml]: clean comment code -- we can leave it until I run some jobs on mast

File: torchtitan/datasets/download_tokenizer.py [hunk 12]
```diff
@@ -38,12 +42,24 @@ def hf_download(repo_id: Optional[str] = None, hf_token: Optional[str] = None) -
...
         type=str,
-        default="meta-llama/llama-2-70b",
-        help="Repository ID to download from.",
+        default="meta-llama/Meta-Llama-3-8B",
+        help="Repository ID to download from. default to Llama-3-8B",
+    )
+    parser.add_argument(
+        "--tokenizer_path",
+        type=str,
+        default="original",
```
Comment (tianyu-l) [inline torchtitan/datasets/download_tokenizer.py]: Setting `default="original"` makes llama2 tokenizer download instruction in README invalid (which requires `--tokenizer_path = ""`).

File: train_configs/llama3_8b.toml [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+[metrics]
+log_freq = 10
+enable_tensorboard = true
+save_tb_folder = "tb"
+
+[model]
+name = "llama3"
+flavor = "8B"
+tokenizer_path = "./torchtitan/datasets/tokenizer/original/tokenizer.model"
```
Comment (tianyu-l) [inline train_configs/llama3_8b.toml]: Putting llama2 tokenizer.model under `torchtitan/datasets/tokenizer/` and llama3 tokenizer.model under `/torchtitan/datasets/tokenizer/original/` is confusing, especially because they share the same file name. Can we organize them better?

File: README.md [hunk 14]
```diff
@@ -47,15 +47,27 @@ Install PyTorch from source or install the latest pytorch nightly, then install
...
+`torchtitan` current supports training Llama3 (8B, 70B), and Llama2 (13B, 70B) out of box. To get started training these models, we need to download a tokenizer.model. Follow the instructions on the official [meta-llama](https://huggingface.co/meta-llama/Meta-Llama-3-8B) repository to ensure you have access to the Llama model weights.
+
+Once you have confirmed access, you can run the following command to download the Llama2/3 tokenizer to your local machine.
+
 ```
-pip install -r dev-requirements.txt
+# pass your hf_token in order to download tokenizer.model
+
+# llama3 tokenizer.model
+python torchtitan/datasets/download_tokenizer.py --repo_id meta-llama/Meta-Llama-3-8B --local_dir "original" --hf_token=...
```
Comment (tianyu-l) [inline README.md]: should 'local_dir' be 'tokenizer_path'?

File: train_configs/llama3_8b.toml [hunk 15]
```diff
@@ -0,0 +1,47 @@
+# torchtitan Config.toml
+[job]
+dump_folder = "./outputs"
+description = "LLaMA 3 8B training"
+
+[profiling]
```
Comment (tianyu-l) [inline train_configs/llama3_8b.toml:6]: llama3_8b.toml seems copied from a outdated version, pls update

File: requirements.txt (hunk 3 above)
Comment (tianyu-l) [inline requirements.txt]: I was asked for `pip install blobfile` when tiktoken is first time running

Comment (tianyu-l) [reply requirements.txt]: needs to use the newest versions 0.5.2 or 0.6.0 to avoid https://github.com/openai/tiktoken/issues/75

Comment (wanchaol) [reply requirements.txt]: Oh I mainly followed the llama3 repo where they fixed the dependency to 0.4.0 (not sure why though)

File: README.md (hunk 5 above)
Comment (wanchaol) [reply README.md:52]: yes! I think we want to shorten the user setup instructions given we add a new step to download tokenizer

File: README.md (hunk 14 above)
Comment (wanchaol) [reply README.md]: good catch!

File: torchtitan/datasets/download_tokenizer.py (hunk 12 above)
Comment (wanchaol) [reply torchtitan/datasets/download_tokenizer.py]: make sense I think I'll keep the default be empty

File: requirements.txt (hunk 3 above)
Comment (wanchaol) [reply requirements.txt]: Is it still the case for tiktokenizer version > 0.5.2? Or we have to install the blob file?

File: torchtitan/datasets/tokenizer/base_tokenizer.py (hunk 4 above)
Comment (wanchaol) [reply torchtitan/datasets/tokenizer/base_tokenizer.py]: Sounds good I think I'll go with `Tokenizer` as the base class

File: torchtitan/models/__init__.py (hunk 8 above)
Comment (wanchaol) [reply torchtitan/models/__init__.py]: we can do that, that would need to change all our existing llama2 configs though, I think it's worth doing that actually though, let me change all names to align

File: requirements.txt (hunk 3 above)
Comment (tianyu-l) [reply requirements.txt]: that's a good question... I haven't tried

on mast it requires an additional pkg `chardet`, which was not asked on local devgpu

Comment (wanchaol) [reply requirements.txt]: hmmm I guess I'll try to remove the version pin for now
